File: 2022/21.jl
include("../utils.jl")
using .AoCUtils: readInput

function loadInput()
    filename = "input_21.txt"
    lines = readInput(filename)
    
    ingredients = String[]
    allergens = String[]
    recipes = Vector{Tuple{String, Vector{String}}}()
    real_recipes = Vector{Vector{String}}()
    for l in lines
        items = split(l, " (contains ")
        ingr = split(items[1])
        allgs = split(items[2][1:end-1], ", ")
        append!(ingredients, ingr)
        append!(allergens, allgs)
        push!(real_recipes, ingr)
        for a in allgs
            push!(recipes, (a, ingr))
        end
    end

    return unique(ingredients), unique(allergens), recipes, real_recipes
end

function part1(ingrs::Vector{String}, allgs::Vector{String},
               recipes::Vector{Tuple{String, Vector{String}}},
               real_recipes::Vector{Vector{String}})
    allgs_ingrs = Dict{String, Set{String}}()
    for a in allgs
        for (a1, i) in recipes
            if a == a1
                if haskey(allgs_ingrs, a)
                    intersect!(allgs_ingrs[a], Set(i))
                else
                    allgs_ingrs[a] = Set(i)
                end
            end
        end
    end
    
    while true
        removed = false
        for (a, i) in allgs_ingrs
            if length(i) == 1
                i = first(i)
                for (a1, i1) in allgs_ingrs
                    if a != a1 && i in i1
                        delete!(allgs_ingrs[a1], i)
                        removed = true
                    end
                end
            end
        end
 
        if !removed
            break
        end
    end

    mapped = [first(i) for i in values(allgs_ingrs)]
    not_mapped = setdiff(Set(ingrs), Set(mapped))

    val = 0
    for r in real_recipes
        for m in not_mapped
            if m in r
                val += 1
            end
        end
    end

    println("🎄 Part 1: $(val)")
    return allgs_ingrs
end

function part2(allgs_ingr::Dict{String, Set{String}})
    ing = [first(allgs_ingr[k]) for k in sort(collect(keys(allgs_ingr)))]
    println("🎄🎅 Part 2: $(join(ing, ","))")
end

println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
println("⛄        Day 21        ⛄")
println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

ingrs, allgs, recipes, real_recipes = loadInput()
@time allgs_ingr = part1(ingrs, allgs, recipes, real_recipes)
@time part2(allgs_ingr)

File: 2022/21.py
from utils import readInput
import time

def part1():
    ingredients = []
    allergens = set([])
    receipts = []
    with open("input_21.txt") as f:
        for i, l in enumerate(f):
            l = l.split("\n")[0]
            items = l.split(" (contains ")
            ingr = items[0].split(" ")
            receipts.append(ingr)
            a = items[1][:-1].split(", ")
            for i in a:
                allergens.add(i)
                ingredients.append((i, ingr))

    all_ingredients = []
    for ing in ingredients:
        all_ingredients += ing[1]

    all_ingredients = set(all_ingredients)
    
    final = {}
    for a in allergens:
        for ing in ingredients:
            if a == ing[0]:
                if a in final:
                    final[a] = final[a].intersection(set(ing[1]))
                else:
                    final[a] = set(ing[1])

    while True:
        removed = False
        for k,v in final.items():
            v = list(v)
            if len(v) == 1:
                for k1 in final:
                    if k != k1:
                        if v[0] in final[k1]:
                            final[k1].remove(v[0])
                            removed = True

        if not removed:
            break

    ok = []
    for v in final.values():
        ok.append(list(v)[0])
    ok = all_ingredients - set(ok)

    val = 0
    for r in receipts:
        for o in ok:
            if o in r:
                val += 1

    print ("🎄 Part 1: {}".format(val))
    return final

def part2(ok):
    ing = [list(ok[k])[0] for k in sorted(ok)]
    print ("🎄🎅 Part 2: {}".format(",".join(ing)))
    
t0 = time.time()
ok = part1()
print ("Time: {:.5f}".format(time.time()-t0))

t0 = time.time()
part2(ok)
print ("Time: {:.5f}".format(time.time()-t0))
